### PR TITLE
[AN-1828] [AN-1944] Feature/v7 billing migration

### PR DESCRIPTION
**Pull request by marcelorbenites**
- Credit Card option for In-Apps.
- PayPal option for Paid Apps and In-Apps.
- In-Apps use V7 Billing API.
- Paid Apps use V3 Billing API.
- V7 Billing API is only in DEV server therefore it only works for variants composed by Internal flavor.

### CODE REVIEW

File: app/src/main/java/cm/aptoide/pt/AptoideApplication.java [hunk 1]
```diff
@@ -745,19 +718,20 @@ public AptoideAccountManager getAccountManager() {
...
 
       final AccountService accountService =
-          new V3AccountService(accountFactory, getDefaultClient(), getLongTimeoutClient(),
+          new AccountServiceV3(accountFactory, getDefaultClient(), getLongTimeoutClient(),
               WebService.getDefaultConverter(), getNonNullObjectMapper(),
               getDefaultSharedPreferences(), getExtraId(), getTokenInvalidator(),
               getAuthenticationPersistence(), getNoAuthenticationBodyInterceptorV3(),
-              getMultipartBodyInterceptor(), getBodyInterceptorWebV7(), getBodyInterceptorPoolV7());
+              getMultipartBodyInterceptor(), getBodyInterceptorPoolV7(),
+              getBodyInterceptorPoolV7());
```
Comment (jdandrade) [inline app/src/main/java/cm/aptoide/pt/AptoideApplication.java]: Two Interceptors of Pool type are being injected, one should be Web?

File: app/src/main/java/cm/aptoide/pt/billing/authorization/AuthorizationFactory.java [hunk 2]
```diff
@@ -1,75 +1,47 @@
...
 
-    if (response.getAuthorizations() != null && response.getAuthorizations()
-        .isEmpty()) {
-      return Authorization.Status.INACTIVE;
+    if (authorization instanceof WebAuthorization) {
+      return WEB;
     }
 
-    return Authorization.Status.UNKNOWN_ERROR;
+    return null;
```
Comment (jdandrade) [inline app/src/main/java/cm/aptoide/pt/billing/authorization/AuthorizationFactory.java]: Question: Shouldn't this return a something else than null? Even though null type is being handled in the create method.

File: app/src/main/java/cm/aptoide/pt/billing/external/ExternalBillingBinder.java [hunk 3]
```diff
@@ -56,25 +60,34 @@
...
   public ExternalBillingBinder(Context context, ExternalBillingSerializer serializer,
-      PaymentThrowableCodeMapper errorCodeFactory, Billing billing, CrashReport crashReport,
-      BillingIdResolver idResolver, int apiVersion, BillingAnalytics analytics) {
+      PaymentThrowableCodeMapper errorCodeFactory, CrashReport crashReport, int apiVersion,
+      BillingAnalytics analytics, PackageManager packageManager) {
     this.context = context;
     this.serializer = serializer;
     this.errorCodeFactory = errorCodeFactory;
+    this.packageManager = packageManager;
     this.billing = billing;
```
Comment (jdandrade) [inline app/src/main/java/cm/aptoide/pt/billing/external/ExternalBillingBinder.java]: Variable not being assigned, but is instantiated in the onTransact

Comment (jdandrade) [inline app/src/main/java/cm/aptoide/pt/billing/external/ExternalBillingBinder.java]: Variable not being assigned, but is instantiated in the onTransact

File: app/src/main/java/cm/aptoide/pt/billing/external/ExternalBillingBinder.java [hunk 4]
```diff
@@ -56,25 +60,34 @@
...
     this.crashReport = crashReport;
-    this.idResolver = idResolver;
     this.supportedApiVersion = apiVersion;
     this.analytics = analytics;
   }
 
+  @Override public boolean onTransact(int code, Parcel data, Parcel reply, int flags)
+      throws RemoteException {
+    merchantName = packageManager.getPackagesForUid(Binder.getCallingUid())[0];
```
Comment (jdandrade) [inline app/src/main/java/cm/aptoide/pt/billing/external/ExternalBillingBinder.java]: Warning: may produce NPE?

File: app/src/main/java/cm/aptoide/pt/AptoideApplication.java (hunk 1 above)
Comment (marcelorbenites) [reply app/src/main/java/cm/aptoide/pt/AptoideApplication.java]: Fixed.

File: app/src/main/java/cm/aptoide/pt/billing/external/ExternalBillingBinder.java (hunk 3 above)
Comment (marcelorbenites) [reply app/src/main/java/cm/aptoide/pt/billing/external/ExternalBillingBinder.java]: Fixed.

File: app/src/main/java/cm/aptoide/pt/billing/networking/TransactionServiceV3.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+public class TransactionServiceV3 implements TransactionService {
+
+  private final TransactionMapperV3 transactionMapper;
+  private final BodyInterceptor<BaseBody> bodyInterceptorV3;
+  private final Converter.Factory converterFactory;
+  private final OkHttpClient httpClient;
+  private final TokenInvalidator tokenInvalidator;
+  private final SharedPreferences sharedPreferences;
+  private final TransactionFactory transactionFactory;
+  private final Customer customer;
```
Comment (jdandrade) [inline app/src/main/java/cm/aptoide/pt/billing/networking/TransactionServiceV3.java]: Variable not being used.

File: app/src/main/java/cm/aptoide/pt/billing/payment/OnSubscribeCreditCardFragment.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
+package cm.aptoide.pt.billing.payment;
+
+import cm.aptoide.pt.navigator.FragmentNavigator;
+import com.adyen.core.PaymentRequest;
+import com.adyen.core.models.paymentdetails.CreditCardPaymentDetails;
+import com.adyen.ui.fragments.CreditCardFragmentBuilder;
+import rx.Observer;
+import rx.observables.SyncOnSubscribe;
+
+public class OnSubscribeCreditCardFragment
```
Comment (jdandrade) [inline app/src/main/java/cm/aptoide/pt/billing/payment/OnSubscribeCreditCardFragment.java]: This class is not being used.

File: app/src/main/java/cm/aptoide/pt/billing/payment/SharedPreferencesPaymentServiceSelector.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+            getService(services, defaultServiceType))
+            .switchIfEmpty(Observable.just(services.get(0)))
+            .first()
+            .toSingle());
+  }
+
+  @Override public Completable selectService(PaymentService service) {
+    return Completable.fromAction(() -> preferences.edit()
+        .putString(SELECTED_SERVICE_TYPE, service.getType())
+        .commit())
```
Comment (jdandrade) [inline app/src/main/java/cm/aptoide/pt/billing/payment/SharedPreferencesPaymentServiceSelector.java]: Consider using apply() instead

File: app/src/main/java/cm/aptoide/pt/billing/external/ExternalBillingBinder.java (hunk 4 above)
Comment (marcelorbenites) [reply app/src/main/java/cm/aptoide/pt/billing/external/ExternalBillingBinder.java]: Will never happen and if it does it should crash in runtime.

File: app/src/main/java/cm/aptoide/pt/billing/networking/TransactionServiceV3.java (hunk 5 above)
Comment (marcelorbenites) [reply app/src/main/java/cm/aptoide/pt/billing/networking/TransactionServiceV3.java]: Fixed.

File: app/src/main/java/cm/aptoide/pt/billing/payment/OnSubscribeCreditCardFragment.java (hunk 6 above)
Comment (marcelorbenites) [reply app/src/main/java/cm/aptoide/pt/billing/payment/OnSubscribeCreditCardFragment.java]: Fixed.

File: app/src/main/java/cm/aptoide/pt/billing/payment/SharedPreferencesPaymentServiceSelector.java (hunk 7 above)
Comment (marcelorbenites) [reply app/src/main/java/cm/aptoide/pt/billing/payment/SharedPreferencesPaymentServiceSelector.java]: The observable is subscribed on I/O scheduler therefore I can execute the commit synchronously. In fact I have to make sure it finishes the operation before the action returns. Using apply I could end up with an inconsistent state.

File: app/src/main/java/cm/aptoide/pt/billing/authorization/AuthorizationFactory.java (hunk 2 above)
Comment (marcelorbenites) [reply app/src/main/java/cm/aptoide/pt/billing/authorization/AuthorizationFactory.java]: Won't fix.

File: app/src/main/java/cm/aptoide/pt/billing/view/BillingNavigator.java [hunk 8]
```diff
@@ -28,47 +31,49 @@
   private final ActivityNavigator activityNavigator;
   private final FragmentNavigator fragmentNavigator;
   private final String marketName;
+  private final PublishRelay<CreditCardPaymentDetails> detailsRelay;
+  private final Adyen adyen;
```
Comment (jdandrade) [inline app/src/main/java/cm/aptoide/pt/billing/view/BillingNavigator.java]: Variable not being used.

**Pull request merged**